File: pi/pi.cpp
// provides rand and RAND_MAX
#include <cstdlib>

#include "pi.h"

double distance_to_origin_squared(const double x, const double y)
{
    return x * x + y * y;
}

double approximate_pi_c(const int num_points)
{
    int num_inside = 0;

    for (int i = 0; i < num_points; i++)
    {
        double x = ((double)rand() / (RAND_MAX));
        double y = ((double)rand() / (RAND_MAX));
        if (distance_to_origin_squared(x, y) < 1.0)
        {
            num_inside++;
        }
    }

    // we multiply by 4 to get the full circle
    // from the 4 segments
    return 4.0 * num_inside / ((double)num_points);
}

File: pi/pi.f90
module pi

    implicit none

    public approximate_pi

    private

contains

    pure real(8) function distance_to_origin_squared(x, y)
        real(8), intent(in) :: x
        real(8), intent(in) :: y

        distance_to_origin_squared = x*x + y*y
    end function

    real(8) function approximate_pi(num_points)
        integer, intent(in) :: num_points

        integer :: i
        integer :: num_inside
        real(8) :: x
        real(8) :: y

        num_inside = 0

        do i = 1, num_points
            call random_number(x)
            call random_number(y)
            if (distance_to_origin_squared(x, y) < 1.0) then
                num_inside = num_inside + 1
            end if
        end do

        ! we multiply by 4 to get the full circle
        ! from the 4 segments
        approximate_pi = 4.0*num_inside/real(num_points)
    end function

end module
